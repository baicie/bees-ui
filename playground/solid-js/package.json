{
  "name": "solid-js",
  "private": true,
  "version": "0.0.0",
  "type": "module",
  "scripts": {
    "start": "vite",
    "build-play": "tsc && vite build",
    "preview": "vite preview"
  },
  "dependencies": {
    "solid-js": "^1.8.7"
  },
  "devDependencies": {
    "typescript": "^5.3.3",
    "vite": "^5.0.12",
    "vite-plugin-solid": "^2.9.1"
  }
}