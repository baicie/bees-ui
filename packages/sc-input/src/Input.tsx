import clsx from 'clsx';
import { createEffect, createSignal, splitProps } from 'solid-js';
import type { JSX } from 'solid-js';

import BaseInput from './BaseInput';
import useCount from './hooks/useCount';
import type { ChangeEventInfo, InputProps } from './interface';
import type { InputFocusOptions } from './utils/commonUtils';
import { resolveOnChange, triggerFocus } from './utils/commonUtils';

const Input = (props: InputProps) => {
  const [local, rest] = splitProps(props, [
    'autoComplete',
    'onChange',
    'onFocus',
    'onBlur',
    'onPressEnter',
    'onKeyDown',
    'onKeyUp',
    'prefixCls',
    'disabled',
    'htmlSize',
    'className',
    'maxLength',
    'suffix',
    'showCount',
    'count',
    'type',
    'classes',
    'classNames',
    'styles',
    'onCompositionStart',
    'onCompositionEnd',
    'hidden',
  ]);

  const prefixCls = local.prefixCls || 'rc-input';
  const [focused, setFocused] = createSignal(false);
  const compositionRef = { current: false };
  const keyLockRef = { current: false };

  let inputRef: HTMLInputElement | HTMLTextAreaElement;
  // let holderRef;

  // ====================== Value =======================
  const [value, setValue] = createSignal(props.value ?? props.defaultValue ?? '');

  const formatValue = value === undefined || value === null ? '' : String(value);

  // =================== Select Range ===================
  const [selection, setSelection] = createSignal(null);

  // ====================== Count =======================
  const countConfig = useCount(local.count, local.showCount);
  const mergedMax = countConfig().max || (local.maxLength as number);
  const valueLength = () => countConfig().strategy(formatValue);

  const isOutOfRange = () => !!mergedMax && valueLength() > mergedMax;

  // ====================== Ref =========================
  const focus = (option?: InputFocusOptions) => {
    if (inputRef) {
      triggerFocus(inputRef, option);
    }
  };

  // const blur = () => {
  //   inputRef?.blur();
  // };

  // Watch for disabled state
  createEffect(() => {
    if (local.disabled && focused()) {
      setFocused(false);
    }
  });

  const triggerChange = (
    e: MouseEvent | Event | CompositionEvent,
    currentValue: string,
    info: ChangeEventInfo,
  ) => {
    let cutValue = currentValue;

    if (
      !compositionRef.current &&
      countConfig().exceedFormatter &&
      countConfig().max &&
      countConfig().strategy(currentValue) > countConfig().max
    ) {
      cutValue = countConfig().exceedFormatter(currentValue, {
        max: countConfig().max,
      });

      if (currentValue !== cutValue) {
        setSelection([inputRef.selectionStart || 0, inputRef.selectionEnd || 0]);
      }
    } else if (info.source === 'compositionEnd') {
      return;
    }
    setValue(cutValue);

    if (inputRef) {
      resolveOnChange(inputRef, e, local.onChange as any, cutValue);
    }
  };

  createEffect(() => {
    if (selection()) {
      inputRef?.setSelectionRange(selection()[0], selection()[1]);
    }
  });

  const onInternalChange: JSX.CustomEventHandlersCamelCase<any>['onChange'] = (e) => {
    triggerChange(e, e.target.value, {
      source: 'change',
    });
  };

  const onInternalCompositionEnd: JSX.DOMAttributes<any>['onCompositionEnd'] = (e) => {
    compositionRef.current = false;
    triggerChange(e, e.currentTarget.value, {
      source: 'compositionEnd',
    });
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    //@ts-expect-error
    local.onCompositionEnd?.(e);
  };

  const handleKeyDown: JSX.CustomEventHandlersCamelCase<any>['onKeyDown'] = (e) => {
    if (local.onPressEnter && e.key === 'Enter' && !keyLockRef.current) {
      keyLockRef.current = true;
      local.onPressEnter(e);
    }
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    //@ts-expect-error
    local.onKeyDown?.(e);
  };

  const handleKeyUp: JSX.CustomEventHandlersCamelCase<any>['onKeyUp'] = (e) => {
    if (e.key === 'Enter') {
      keyLockRef.current = false;
    }
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    //@ts-expect-error
    local.onKeyUp?.(e);
  };

  const handleFocus: JSX.CustomEventHandlersCamelCase<any>['onFocus'] = (e) => {
    setFocused(true);
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    //@ts-expect-error
    local.onFocus?.(e);
  };

  const handleBlur: JSX.CustomEventHandlersCamelCase<any>['onBlur'] = (e) => {
    setFocused(false);
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    //@ts-expect-error
    local.onBlur?.(e);
  };

  const handleReset = (e: CompositionEvent | Event | MouseEvent) => {
    setValue('');
    focus();
    resolveOnChange(inputRef, e, local.onChange as any);
  };

  const getInputElement = () => {
    return (
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-expect-error
      <input
        autocomplete={local.autoComplete}
        {...rest}
        onChange={onInternalChange}
        onFocus={handleFocus}
        onBlur={handleBlur}
        onKeyDown={handleKeyDown}
        onKeyUp={handleKeyUp}
        class={clsx(
          prefixCls,
          { [`${prefixCls}-disabled`]: local.disabled },
          local.classNames?.input,
        )}
        style={local.styles?.input}
        ref={inputRef as any}
        size={local.htmlSize}
        type={local.type || 'text'}
        onCompositionStart={(e) => {
          compositionRef.current = true;
          // eslint-disable-next-line @typescript-eslint/ban-ts-comment
          // @ts-expect-error
          local.onCompositionStart?.(e);
        }}
        onCompositionEnd={onInternalCompositionEnd}
      />
    );
  };

  const getSuffix = () => {
    const hasMaxLength = Number(mergedMax) > 0;

    if (local.suffix || countConfig().show) {
      const dataCount = countConfig().showFormatter
        ? countConfig().showFormatter({
            value: formatValue,
            count: valueLength(),
            maxLength: mergedMax,
          })
        : `${valueLength()}${hasMaxLength ? ` / ${mergedMax}` : ''}`;

      return (
        <>
          {countConfig().show && (
            <span
              class={clsx(
                `${prefixCls}-show-count-suffix`,
                { [`${prefixCls}-show-count-has-suffix`]: !!local.suffix },
                local.classNames?.count,
              )}
              style={local.styles?.count}
            >
              {dataCount}
            </span>
          )}
          {local.suffix}
        </>
      );
    }
    return null;
  };

  return (
    <BaseInput
      {...rest}
      prefixCls={prefixCls}
      className={clsx(local.className, isOutOfRange() && `${prefixCls}-out-of-range`)}
      handleReset={handleReset}
      value={formatValue}
      focused={focused()}
      triggerFocus={focus}
      suffix={getSuffix()}
      disabled={local.disabled}
      classes={local.classes}
      classNames={local.classNames}
      styles={local.styles}
      hidden={local.hidden}
    >
      {getInputElement()}
    </BaseInput>
  );
};

export default Input;
