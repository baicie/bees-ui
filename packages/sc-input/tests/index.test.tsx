import { fireEvent, render } from '@bees-ui/testing-library';
import { createSignal } from 'solid-js';
import { describe, expect, it, vi } from 'vitest';

import Input from '../src';

describe('Input', () => {
  it('should support maxLength', () => {
    const [value, setValue] = createSignal('');
    const { container } = render(() => (
      <Input
        prefixCls="rc-input"
        maxLength={3}
        value={value()}
        onInput={(e) => setValue(e.currentTarget.value)}
      />
    ));
    expect(container).toMatchSnapshot();
  });

  it('not crash when value is number', () => {
    const [value, setValue] = createSignal(1);
    const { container } = render(() => (
      <Input
        suffix="Bamboo"
        value={value()}
        onInput={(e) => setValue(Number(e.currentTarget.value))}
      />
    ));
    expect(container).toBeTruthy();
  });

  it('should trigger onFocus and onBlur', () => {
    const onFocus = vi.fn();
    const onBlur = vi.fn();
    const { container } = render(() => <Input onFocus={onFocus} onBlur={onBlur} />);
    const inputEl = container.querySelector('input')!;
    fireEvent.focus(inputEl);
    expect(onFocus).toHaveBeenCalled();
    fireEvent.blur(inputEl);
    expect(onBlur).toHaveBeenCalled();
  });

  it('should trigger onKeydown and onPressEnter', () => {
    const onKeyDown = vi.fn();
    const onPressEnter = vi.fn();
    const { container } = render(() => <Input onKeyDown={onKeyDown} onPressEnter={onPressEnter} />);
    const inputEl = container.querySelector('input')!;
    fireEvent.keyDown(inputEl, { key: 'Enter' });
    expect(onKeyDown).toHaveBeenCalled();
    expect(onPressEnter).toHaveBeenCalled();
  });

  it('should prevent long press of enter', () => {
    const onKeyDown = vi.fn();
    const onPressEnter = vi.fn();
    const onKeyUp = vi.fn();
    const { container } = render(() => (
      <Input onKeyDown={onKeyDown} onPressEnter={onPressEnter} onKeyUp={onKeyUp} />
    ));
    const inputEl = container.querySelector('input')!;
    fireEvent.keyDown(inputEl, { key: 'Enter' });
    fireEvent.keyDown(inputEl, { key: 'Enter' });
    fireEvent.keyUp(inputEl, { key: 'Enter' });
    expect(onKeyDown).toBeCalledTimes(2);
    expect(onPressEnter).toBeCalledTimes(1);
    expect(onKeyUp).toBeCalledTimes(1);
  });

  // it('should trigger onChange', () => {
  //   const onChange = vi.fn();
  //   const [value, setValue] = createSignal('');
  //   const { container } = render(() => (
  //     <Input
  //       onInput={(e) => {
  //         setValue(e.currentTarget.value);
  //         onChange(e);
  //       }}
  //       value={value()}
  //     />
  //   ));
  //   const inputEl = container.querySelector('input')!;
  //   fireEvent.input(inputEl, { target: { value: 'test' } });
  //   expect(onChange).toHaveBeenCalled();
  //   expect(inputEl.value).toBe('test');
  // });

  // it('not block input when `value` is undefined', () => {
  //   const [value, setValue] = createSignal('');
  //   const { container, renderer } = render(() => (
  //     <Input value={value()} onInput={(e) => setValue(e.currentTarget.value)} />
  //   ));
  //   const inputEl = container.querySelector('input')!;
  //   fireEvent.input(inputEl, { target: { value: 'Bamboo' } });
  //   expect(inputEl.value).toEqual('Bamboo');

  //   // Controlled
  //   renderer(() => <Input value="Light" onInput={(e) => setValue(e.currentTarget.value)} />);
  //   fireEvent.input(inputEl, { target: { value: 'Bamboo' } });
  //   expect(inputEl.value).toEqual('Light');

  //   // Uncontrolled
  //   renderer(() => <Input value={undefined} onInput={(e) => setValue(e.currentTarget.value)} />);
  //   expect(inputEl.value).toEqual('');
  //   fireEvent.input(inputEl, { target: { value: 'Bamboo' } });
  //   expect(inputEl.value).toEqual('Bamboo');
  // });

  it('should focus input after clear', async () => {
    const { container } = render(() => (
      <Input prefixCls="rc-input" allowClear defaultValue="111" />
    ));
    const clearIcon = container.querySelector('.rc-input-clear-icon')!;
    await fireEvent.click(clearIcon);
    expect(document.activeElement).toBe(container.querySelector('input'));
  });

  it('support bigint type', () => {
    // TOFIX: bigint not work in test
    const { container } = render(() => <Input value={Number('2222')} />);
    expect(container).toBeTruthy();
  });

  // it('should be compatible with type="file"', () => {
  //   const onChange = vi.fn();
  //   const { container } = render(() => <Input type="file" onInput={onChange} />);
  //   const inputEl = container.querySelector('input')!;
  //   const file = new File(['(⌐□_□)'], 'file.xml', { type: 'application/xml' });
  //   // upload the file by updating the value attribute of the input
  //   fireEvent.change(inputEl, {
  //     target: { files: [file] },
  //   });

  //   expect(onChange).toHaveBeenCalled();
  //   expect(onChange.mock.calls[0][0].target.files[0]).toBe(file);
  // });
});

describe('should support showCount', () => {
  it('maxLength', () => {
    const [value, setValue] = createSignal('12345');
    const { container } = render(() => (
      <Input
        prefixCls="rc-input"
        maxLength={5}
        showCount
        value={value()}
        onInput={(e) => setValue(e.currentTarget.value)}
      />
    ));
    expect(container.querySelector('input')?.value).toBe('12345');
    expect(container.querySelector('.rc-input-show-count-suffix')?.innerHTML).toBe('5 / 5');
  });

  it('control exceed maxLength', () => {
    const [value, setValue] = createSignal('12345678');
    const { container } = render(() => (
      <Input
        prefixCls="rc-input"
        maxLength={5}
        showCount
        value={value()}
        onInput={(e) => setValue(e.currentTarget.value)}
      />
    ));
    expect(container.querySelector('input')?.value).toBe('12345678');
    expect(container.querySelector('.rc-input-show-count-suffix')?.innerHTML).toBe('8 / 5');
  });

  it('count formatter', () => {
    const [value, setValue] = createSignal('12345');
    const { container } = render(() => (
      <Input
        prefixCls="rc-input"
        maxLength={5}
        showCount={{
          formatter: ({ value, count, maxLength }) => `${value}, ${count}, ${maxLength}`,
        }}
        value={value()}
        onInput={(e) => setValue(e.currentTarget.value)}
      />
    ));
    expect(container.querySelector('input')?.value).toBe('12345');
    expect(container.querySelector('.rc-input-show-count-suffix')?.innerHTML).toBe('12345, 5, 5');
  });
});

describe('Input allowClear', () => {
  it('should change type when click', () => {
    const [value, setValue] = createSignal('111');
    const { container } = render(() => (
      <Input
        prefixCls="rc-input"
        allowClear
        value={value()}
        onInput={(e) => setValue(e.currentTarget.value)}
      />
    ));
    const inputEl = container.querySelector('input')!;
    fireEvent.input(inputEl, { target: { value: '111' } });
    expect(inputEl.value).toEqual('111');
    expect(container).toMatchSnapshot();
    fireEvent.click(container.querySelector('.rc-input-clear-icon')!);
    expect(container).toMatchSnapshot();
    expect(inputEl.value).toEqual('');
  });

  it('should not show icon if value is undefined or empty string', () => {
    const containers = [undefined, ''].map(
      (val) => render(() => <Input prefixCls="rc-input" allowClear value={val} />).container,
    );
    containers.forEach((container) => {
      expect(container.querySelector('input')?.value).toEqual('');
      expect(container.querySelector('.rc-input-clear-icon-hidden')).toBeTruthy();
      expect(container).toMatchSnapshot();
    });
  });

  it('should not show icon if defaultValue is undefined or empty string', () => {
    const containers = [undefined, ''].map(
      (val) => render(() => <Input prefixCls="rc-input" allowClear defaultValue={val} />).container,
    );
    containers.forEach((container) => {
      expect(container.querySelector('input')?.value).toEqual('');
      expect(container.querySelector('.rc-input-clear-icon-hidden')).toBeTruthy();
      expect(container).toMatchSnapshot();
    });
  });

  // it('should trigger event correctly', () => {
  //   let argumentEventObject;
  //   let argumentEventObjectValue;
  //   const onChange = (e: Event) => {
  //     argumentEventObject = e;
  //     argumentEventObjectValue = (e.target as HTMLInputElement).value;
  //   };
  //   const { container } = render(() => (
  //     <Input prefixCls="rc-input" allowClear defaultValue="111" onInput={onChange} />
  //   ));
  //   fireEvent.click(container.querySelector('.rc-input-clear-icon')!);
  //   expect(argumentEventObject).toHaveProperty('type', 'click');
  //   expect(argumentEventObjectValue).toBe('');
  //   expect(container.querySelector('input')?.value).toBe('');
  // });

  // it('should trigger event correctly on controlled mode', () => {
  //   let argumentEventObject;
  //   let argumentEventObjectValue;
  //   const onChange = (e: Event) => {
  //     argumentEventObject = e;
  //     argumentEventObjectValue = (e.target as HTMLInputElement).value;
  //   };
  //   const [value, setValue] = createSignal('111');
  //   const { container } = render(() => (
  //     <Input
  //       prefixCls="rc-input"
  //       allowClear
  //       value={value()}
  //       onInput={(e) => {
  //         setValue(e.currentTarget.value);
  //         onChange(e);
  //       }}
  //     />
  //   ));
  //   fireEvent.click(container.querySelector('.rc-input-clear-icon')!);
  //   expect(argumentEventObject).toHaveProperty('type', 'click');
  //   expect(argumentEventObjectValue).toBe('');
  //   expect(container.querySelector('input')?.value).toBe('');
  // });
});
