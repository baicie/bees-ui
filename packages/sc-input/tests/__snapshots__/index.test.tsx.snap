// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Input > should support maxLength 1`] = `
<div>
  <input
    class="rc-input"
    type="text"
  />
</div>
`;

exports[`Input allowClear > should change type when click 1`] = `
<div>
  <span
    class="rc-input-affix-wrapper"
  >
    <input
      allowclear="true"
      class="rc-input"
      type="text"
    />
    <span
      class="rc-input-suffix"
    >
      <span
        class="rc-input-clear-icon"
        role="button"
        tabindex="-1"
      >
        ✖
      </span>
      
    </span>
  </span>
</div>
`;

exports[`Input allowClear > should change type when click 2`] = `
<div>
  <span
    class="rc-input-affix-wrapper rc-input-focused"
  >
    <input
      allowclear="true"
      class="rc-input"
      type="text"
    />
    <span
      class="rc-input-suffix"
    >
      <span
        class="rc-input-clear-icon"
        role="button"
        tabindex="-1"
      >
        ✖
      </span>
      
    </span>
  </span>
</div>
`;

exports[`Input allowClear > should not show icon if defaultValue is undefined or empty string 1`] = `
<div>
  <span
    class="rc-input-affix-wrapper"
  >
    <input
      allowclear="true"
      class="rc-input"
      type="text"
    />
    <span
      class="rc-input-suffix"
    >
      <span
        class="rc-input-clear-icon rc-input-clear-icon-hidden"
        role="button"
        tabindex="-1"
      >
        ✖
      </span>
      
    </span>
  </span>
</div>
`;

exports[`Input allowClear > should not show icon if defaultValue is undefined or empty string 2`] = `
<div>
  <span
    class="rc-input-affix-wrapper"
  >
    <input
      allowclear="true"
      class="rc-input"
      defaultvalue=""
      type="text"
    />
    <span
      class="rc-input-suffix"
    >
      <span
        class="rc-input-clear-icon rc-input-clear-icon-hidden"
        role="button"
        tabindex="-1"
      >
        ✖
      </span>
      
    </span>
  </span>
</div>
`;

exports[`Input allowClear > should not show icon if value is undefined or empty string 1`] = `
<div>
  <span
    class="rc-input-affix-wrapper"
  >
    <input
      allowclear="true"
      class="rc-input"
      type="text"
    />
    <span
      class="rc-input-suffix"
    >
      <span
        class="rc-input-clear-icon rc-input-clear-icon-hidden"
        role="button"
        tabindex="-1"
      >
        ✖
      </span>
      
    </span>
  </span>
</div>
`;

exports[`Input allowClear > should not show icon if value is undefined or empty string 2`] = `
<div>
  <span
    class="rc-input-affix-wrapper"
  >
    <input
      allowclear="true"
      class="rc-input"
      type="text"
    />
    <span
      class="rc-input-suffix"
    >
      <span
        class="rc-input-clear-icon rc-input-clear-icon-hidden"
        role="button"
        tabindex="-1"
      >
        ✖
      </span>
      
    </span>
  </span>
</div>
`;
